working/0110-incremental-sound-nnbd/feature-specification.md

```diff
@@ -0,0 +1,374 @@
+before this feature ships (or the last version to which it has opted in, if that
+is different).  All opted-in libraries upstream from the legacy library are
+viewed by the legacy library with nullability related features erased from their
+APIs.  In particular:
+  - All types of the form `T?` in the opted-in API are treated as `T`.
+  - All required named parameters are treated as optional named parameters.
+  - The type `Never` is treated as the type `Null`
+
+In a legacy library, none of the new syntax introduced by this proposal is
+available, and it is a static error if it is used.
+
+### Importing legacy libraries from opted-in libraries
+
+The type system is extended with a notion of a legacy type operator.  For every
+type `T`, there is an additional type `T*` which is the legacy version of the
+type.  There is no surface syntax for legacy types, and implementations should
+display the legacy type `T*` in the same way that they would display the type
+`T`, except in so far as it is useful to communicate to programmers for the
+purposes of error messages that the type originates in unmigrated code.
+
+When static checking is done in a migrated library, types which are imported
+from unmigrated libraries are seen as legacy types.  However, for the purposes
+of type inference in migrated libraries, types imported from unmigrated
+libraries shall be treated as non-nullable.  As a result, legacy types will
+never appear as type annotations in migrated libraries, nor will they appear in
+reified positions.
+
+### Type reification
+
+All types reified in legacy libraries are reified as legacy types.  Runtime
+subtyping checks treat them according to the subtyping rules specified
+separately.
+
+### Runtime checks and weak checking
+
+When weak checking is enabled, runtime type tests (including explicit and
+implicit casts) shall succeed with a warning whenever the runtime type test
+would have succeeded if all `?` types were ignored, `Never` were treated as
+`Null`, and `required` named parameters were treated as optional.
+
+### Exports
+
+If an unmigrated library re-exports a migrated library, the re-exported symbols
+retain their migrated status (that is, downstream migrated libraries will see
+their migrated types).
+
+It is an error for a migrated library to re-export symbols from an unmigrated
+library.
+
+### Override checking
+
+In an unmigrated library, override checking is done using legacy types.  This
+means that an unmigrated library can bring together otherwise incompatible
+methods.  When choosing the most specific signature during interface
+computation, all nullability and requiredness annotations are ignored, and the
+`Never` type is treated as `Null`.
+
+In a migrated library, override checking must check that an override is
+consistent with all overridden methods from other migrated libraries in the
+super-interface chain, since a legacy library is permitted to override otherwise
+incompatible signatures for a method.
+
```

Comment:
(What do we do if a migrated class extends an unmigrated class, and does not override an umigrated method? Will the type in the subclass interface be the `*`-erasure of the original type? Will it be enforced?)

Reply:
I think that inherited methods should stay as legacy types.  So:
```
// @dart=2.3
library foo
class A {
  int foo(int x) => null;
}

class B extends A {
}

void test() {
  int x = foo(null); // Ok
}
 ```

Reply:
That's probably the only sane choice. If you don't write anything, we can't guess how to migrate the legacy function, so we just have to keep it at the legacy signature.